### PR TITLE
feat: enable feature grouping for attention mechanism

### LINKED ISSUE #122: Research : Embedding Aware Attention
# Main Problem

When training with large embedding dimensions, the mask size goes up.

One problem I see is that sparsemax does not know about which columns come from the same embedded columns, this could create something a bit difficult for the model to learn: 
- create embeddings that make sense
- mask embeddings without destroying them, in fact since sparsemax is sparse it's very unlikely that all the columns from a same embedding are used, so you lose the power of your embedding

# Proposed Solutions

It's an open problem but one way I see as promising is to create embedding aware attention.

 The idea would be to mask all dimensions from a same embedding the same way, either by using the mean or the max of the initial mask.

I implemented a first version here : #92 

If you feel like this is interesting and would like to contribute, please share your ideas in comments or open a PR!

### PR DESCRIPTION
**Pull request by Optimox**
**What kind of change does this PR introduce?**
This PR solves #122 in a new way.

Embeddings from a same column are automatically grouped together.
Now attention works at the group level and not feature level. So without specifying anything groups for different categorical features are created.

This is a new as it allows users to specify features they would like to be grouped together by the attention mechanism. This can be very useful when sparse features are created (for example after a TD-IDF), attention has a hard time using this kind of features because of sparsity coming from both the data and the attention. Now you can group all those features together and have a single attention.

**Does this PR introduce a breaking change?**
I'm not sure if this should be considered a breaking change or not. I think so as old trained models used with this new code will have a different behaviour.

**What needs to be documented once your changes are merged?**

I think all changes are already documented in this PR.

**Closing issues**
closes #122

### CODE REVIEW

File: pytorch_tabnet/tab_network.py [hunk 1]
```diff
@@ -176,24 +189,25 @@ def forward(self, x, prior=None):
...
+        bs = x.shape[0]  # batch size
+        prior = torch.ones((bs, self.attention_dim)).to(x.device)
         M_explain = torch.zeros(x.shape).to(x.device)
         att = self.initial_splitter(x)[:, self.n_d :]
         masks = {}
 
         for step in range(self.n_steps):
             M = self.att_transformers[step](prior, att)
-            masks[step] = M
+            M_feature_level = x = torch.matmul(M, self.group_attention_matrix)
```
Comment (eduardocarvp) [inline pytorch_tabnet/tab_network.py:200]: Why redefine `x` here? Are we not losing the input information?

File: pytorch_tabnet/utils.py [hunk 2]
```diff
@@ -216,6 +216,93 @@ def create_explain_matrix(input_dim, cat_emb_dim, cat_idxs, post_embed_dim):
...
+def create_group_matrix(list_groups, input_dim):
+    """
+    Create the group matrix corresponding to the given list_groups
+
+    Parameters
+    ----------
+    - list_groups : list of list of int
+        Each element is a list representing features in the same group.
+        One feature should appear in maximum one group.
+        Feature that don't get assign a group will be in their own group of one feature.
```
Comment (eduardocarvp) [inline pytorch_tabnet/utils.py:228]: assign -> assigned

File: pytorch_tabnet/utils.py [hunk 3]
```diff
@@ -216,6 +216,93 @@ def create_explain_matrix(input_dim, cat_emb_dim, cat_idxs, post_embed_dim):
...
+        - is a list of list
+        - does not contain twice the same feature in different groups
+        - does not contain unknown features (>= input_dim)
+        - does not contain empty groups
+    Parameters
+    ----------
+    - list_groups : list of list of int
+        Each element is a list representing features in the same group.
+        One feature should appear in maximum one group.
+        Feature that don't get assign a group will be in their own group of one feature.
```
Comment (eduardocarvp) [inline pytorch_tabnet/utils.py:279]: Features that don't get assigned

File: pytorch_tabnet/utils.py [hunk 4]
```diff
@@ -216,6 +216,93 @@ def create_explain_matrix(input_dim, cat_emb_dim, cat_idxs, post_embed_dim):
...
+    """
+    assert isinstance(list_groups, list), "list_groups must be a list of list."
+
+    if len(list_groups) == 0:
+        return
+    else:
+        for group_pos, group in enumerate(list_groups):
+            msg = f"Groups must be given as a list of list, but found {group} in position {group_pos}."  # noqa
+            assert isinstance(group, list), msg
+            assert len(group) > 0, "Empty groups are forbidding please remove empty groups []"
```
Comment (eduardocarvp) [inline pytorch_tabnet/utils.py:290]: forbidding -> forbidden

File: pytorch_tabnet/tab_network.py [hunk 5]
```diff
@@ -176,24 +189,25 @@ def forward(self, x, prior=None):
...
             prior = torch.mul(self.gamma - M, prior)
             # output
-            masked_x = torch.mul(M, x)
+            masked_x = torch.mul(M_feature_level, x)
             out = self.feat_transformers[step](masked_x)
             d = ReLU()(out[:, : self.n_d])
             # explain
             step_importance = torch.sum(d, dim=1)
-            M_explain += torch.mul(M, step_importance.unsqueeze(dim=1))
+            M_explain += torch.mul(M_feature_level, step_importance.unsqueeze(dim=1))
```
[the remark this answers is not in the document]
Comment (Optimox) [reply pytorch_tabnet/tab_network.py:210]: There is already a mapping to get importance from post embedding dimension to initial features. They might be redundant indeed, but I don't think I have time to think about it.

**Review by eduardocarvp — APPROVED** [review]

**Comment by gauravbrills** [comment]
can we add an example of how to use this , bit confused about this @Optimox

**Comment by Optimox** [comment]
@gauravbrills There is an example on this notebook : https://github.com/dreamquark-ai/tabnet/blob/develop/census_example.ipynb

You simply need to give a list of groups, each group is a list with the index of the features forming the group.
Attention mechanism will consider each group as one single feature, so all features of the group will get the same attention (and importance). Note that all embedding dimensions generated by a categorical feature will be grouped together.
